src/librustc_mir/transform/mod.rs

```diff
@@ -267,54 +269,81 @@ fn run_optimization_passes<'tcx>(
     def_id: DefId,
     promoted: Option<Promoted>,
 ) {
+    let post_borrowck_cleanup: &[&dyn MirPass<'tcx>] = &[
+        // Remove all things only needed by analysis
+        &no_landing_pads::NoLandingPads::new(tcx),
+        &simplify_branches::SimplifyBranches::new("initial"),
+        &remove_noop_landing_pads::RemoveNoopLandingPads,
+        &cleanup_post_borrowck::CleanupNonCodegenStatements,
+        &simplify::SimplifyCfg::new("early-opt"),
+        // These next passes must be executed together
+        &add_call_guards::CriticalCallEdges,
+        &elaborate_drops::ElaborateDrops,
+        &no_landing_pads::NoLandingPads::new(tcx),
+        // AddMovesForPackedDrops needs to run after drop
+        // elaboration.
+        &add_moves_for_packed_drops::AddMovesForPackedDrops,
+        // `AddRetag` needs to run after `ElaborateDrops`. Otherwise it should run fairly late,
+        // but before optimizations begin.
+        &add_retag::AddRetag,
+        &simplify::SimplifyCfg::new("elaborate-drops"),
+        // No lifetime analysis based on borrowing can be done from here on out.
+    ];
+
+    let optimizations: &[&dyn MirPass<'tcx>] = &[
+        &unreachable_prop::UnreachablePropagation,
+        &uninhabited_enum_branching::UninhabitedEnumBranching,
+        &simplify::SimplifyCfg::new("after-uninhabited-enum-branching"),
+        &inline::Inline,
+        // Lowering generator control-flow and variables has to happen before we do anything else
+        // to them. We do this inside the "optimizations" block so that it can benefit from
+        // optimizations that run before, that might be harder to do on the state machine than MIR
+        // with async primitives.
+        &generator::StateTransform,
+        &instcombine::InstCombine,
+        &const_prop::ConstProp,
```

Comment:
`ConstProp` contains a test for `mir_opt_level`, should that also be removed?

Reply:
Hmm.... I guess we should either remove the check since we don't run it if optimizations are disabled or we should add it to `no_optimizations` leaving the check. I think the second option would be the same as the current behavior, so I'm going to do that.

Reply:
Okay. Please leave a FIXME though, to evaluate if this actually doing anything with optimizations turned off. (That pass also does some linting, but then CI is passing in your PR so it seems we already don't get those lints... or we just never test `mir-opt-level=0`?)

Reply:
We never test `mir-opt-level=0` and `mir-opt-level=2` is only tested by the `mir-opt` tests.

Reply:
> We never test mir-opt-level=0

That is rather concerning, in particular considering that `StateTransform` might rely on some of the simplifications that are now being skipped having happened (and https://github.com/rust-lang/rust/pull/70073#discussion_r404424193).

(Miri sets `mir-opt-level=0` though, so those code paths are not *entirely* untested.)

Reply:
Maybe it would make sense to take some of the basic `async fn` smoketests in the rustc test suite, and use the revision system to also run them with `mir-opt-level=0`?

Reply:
I think that should look something like
```
// revisions: default nomiropt
//[nomiropt]compile-flags: -Z mir-opt-level=0
```
I usually add `-Zdoesntexist` as well for a first test just to make sure the flag is actually used.

Reply:
It's definitely not ideal. A compiler bootstrap with `mir-opt-level=0` does complete successfully though so it's not *completely* broken.